### ISSUE TITLE
Format hint in `image_info->magick` doesn't work when loading from file

### ISSUE THREAD

**Issue by kleisauke** [ImageMagick/ImageMagick#6242]
### ImageMagick version

7.1.1-6

### Operating system

Linux

### Operating system, version and so on

Fedora 37

### Description

Hinting the format in `image_info->magick` doesn't work when loading from a file. For example, when loading a NEF image and initializing `image_info->magick` to `"NEF"` the image appears to load via the TIFF loader (see reproducer below).

Reverting commit 6f7cebf3fceb41ada062fb782a259db82644dc39 seems to fix this.
```diff
--- a/MagickCore/image.c
+++ b/MagickCore/image.c
@@ -2855,8 +2855,7 @@ MagickExport MagickBooleanType SetImageInfo(ImageInfo *image_info,
         }
     }
   *component='\0';
-  if (*image_info->magick == '\0')
-    GetPathComponent(image_info->filename,ExtensionPath,component);
+  GetPathComponent(image_info->filename,ExtensionPath,component);
   if (*component != '\0')
     {
       /*
```
(This corresponds to commit https://github.com/ImageMagick/ImageMagick6/commit/2cf7bfe4fbcd51025467934b4bccb50511654509 in ImageMagick 6.x)

### Steps to Reproduce

```c
/* Compile with:
 * gcc -g -Wall test-nef.c `pkg-config MagickCore --cflags --libs` -o test-nef
 *
 * Run with:
 * ./test-nef _DSC4175.NEF
 *
 * Expected output: 4284x2844
 * Actual output: 160x120
 */
#include <stdio.h>
#include <assert.h>

#include <MagickCore/MagickCore.h>

int
main(int argc, char **argv)
{
	ExceptionInfo
		*exception;

	Image
		*image;

	ImageInfo
		*image_info;

	if (argc != 2) {
		printf("Usage: %s image\n", argv[0]);
		return 1;
	}

	MagickCoreGenesis(*argv, MagickTrue);
	image_info = AcquireImageInfo();
	(void) CopyMagickString(image_info->filename, argv[1], MaxTextExtent);
	(void) CopyMagickString(image_info->magick, "NEF", MaxTextExtent);
	exception = AcquireExceptionInfo();
	image = ReadImage(image_info, exception);

	assert(image);
	printf("%ldx%ld\n", image->columns, image->rows);

	image = DestroyImage(image);
	exception = DestroyExceptionInfo(exception);
	image_info = DestroyImageInfo(image_info);
	MagickCoreTerminus();

	return 0;
}
```

### Images

[_DSC4175.zip](https://github.com/ImageMagick/ImageMagick/files/11183065/_DSC4175.zip)

**Comment by urban-warrior** [comment]
Thanks for the problem report.  We can reproduce it and will have a patch to fix it in the GIT main branch @ https://github.com/ImageMagick/ImageMagick later today.  The patch will be available in the beta releases of ImageMagick @ https://imagemagick.org/archive/beta/ by sometime tomorrow.

**Comment by kleisauke** [comment]
Thanks! I'm not sure if this is on purpose, but I noticed that `image_info->filename` has a higher priority than `image_info->magick`, i.e. on the above reproducer:
```console
$ cp _DSC4175.NEF x.tiff
$ ./test-nef x.tiff
160x120
$ ./test-nef _DSC4175.NEF
4284x2844
```

I would expect both commands to return `4284x2844`.

**Comment by kleisauke** [comment]
Ah, it looks like `image_info->magick` is still being ignored somehow.
```console
$ cp _DSC4175.NEF x
$ ./test-nef x
160x120
```

**Comment by urban-warrior** [comment]
Will revert and take a closer look at the issue.

**Comment by urban-warrior** [comment]
The type of an image can be implied by its file extension, such as NEF in this case. However, the internal "magic" numbers of the image can override this implicit format. In the case where an image has a TIFF signature, it will be read by the TIFF coder instead. If you want to explicitly set the image type, you can do so by prefixing the image filename. This explicit file type will take precedence over any magic number or implicit file extension.

It's possible that your unit test used to work in the past, but we likely discovered and fixed a bug that caused a side effect, leading to your unit test failure. To achieve the desired outcome, it's best to use an explicit filename.

```
$ ./test-nef nef:_DSC4175.NEF
4284x2844
```

**Comment by kleisauke** [comment]
I see, but NEF images are registered as `ExplicitFormatType`.
https://github.com/ImageMagick/ImageMagick/blob/e59034371872753156cfaa5aea8b63d42121316a/coders/dng.c#L805

So, I would assume that setting `image_info->magick` to `"NEF"` would _not_ determine the image format from the first few bytes of the file, thus preventing it from being loaded by the TIFF coder.

How about this patch?:
```diff
--- a/magick/image.c
+++ b/magick/image.c
@@ -2914,6 +2914,9 @@ MagickExport MagickBooleanType SetImageInfo(ImageInfo *image_info,
     {
       (void) CopyMagickString(magic,image_info->magick,MaxTextExtent);
       magick_info=GetMagickInfo(magic,sans_exception);
+      if ((magick_info != (const MagickInfo *) NULL) &&
+          (magick_info->format_type == ExplicitFormatType))
+          image_info->affirm=MagickTrue;
       if (frames == 0)
         GetPathComponent(image_info->filename,CanonicalPath,filename);
       else
```

This would fix all the test cases mentioned earlier.

**Comment by urban-warrior** [comment]
We'll apply the patch.  We'll need to run all our units tests (multi-thousands) before we can release the patch.  If we run into any issues, we'll let you know.

**Comment by kleisauke** [comment]
Thanks! I'll close, feel free to re-open if this causes any issues.

For reference, this fix was released in version [7.1.1-7](https://github.com/ImageMagick/ImageMagick/releases/tag/7.1.1-7) and [6.9.12-85](https://github.com/ImageMagick/ImageMagick6/releases/tag/6.9.12-85).
